### PR TITLE
Refactor area weighted regridding, improve performance

### PR DESCRIPTION
**Pull request by stephenworsley**
### Description
Implements #5365.

### CODE REVIEW

**Comment by stephenworsley** [comment]
Note: this PR builds a wights matrix and performs regridding with the expectation that the standard cube has the last two dimensions as the y and x dimension (see the change made in this commit https://github.com/SciTools/iris/pull/5543/commits/90130ccc054b336339efeb37458fcaeaf566aa39). This decision is made in order to maintain consistency with patterns established in existing regridders, e.g. the ordering of `dims` expected in this line for example:
https://github.com/SciTools/iris/blob/e0850ebc73824c929f48d51067f88c3c2067b7fb/lib/iris/analysis/_regrid.py#L365
This cosistency is important so that further unification of regridding can take place without having to change the expected ordering of weights matrices associated with regridders.

This decision is also in keeping with recomendations in CF conventions, see http://cfconventions.org/Data/cf-conventions/cf-conventions-1.10/cf-conventions.html#dimensions

File: lib/iris/analysis/_area_weighted.py [hunk 1]
```diff
@@ -1150,3 +459,266 @@ def regrid_callback(*args, **kwargs):
...
+    return result
+
+
+def _regrid_along_dims(data, x_dim, y_dim, weights, tgt_shape, mdtol):
+    """Regrid data, handling masks and dimensions."""
+    # Handle scalar coordinates.
+    # Note: scalar source coordinates are only handled when their
+    # corresponding target coordinate is also scalar.
+    if x_dim is None:
+        x_none = True
```
Comment (HGWright) [inline lib/iris/analysis/_area_weighted.py:699]: Can simplify this code by removing `x_none` and `y_none`, and instead just tracking the number of added scalar coords. This also effects the `np.squeeze` at the bottom of this code.

File: lib/iris/analysis/_area_weighted.py [hunk 2]
```diff
@@ -1150,3 +459,266 @@ def regrid_callback(*args, **kwargs):
...
+            oob_mask = inbound_sums > 1 - 1e-8
+        else:
+            oob_mask = inbound_sums > 1 - mdtol
+        # Broadcast the mask to the shape of the full array
+        oob_slice = ((np.newaxis,) * len(data.shape[:-2])) + np.s_[:, :]
+        tgt_mask = tgt_mask * oob_mask[oob_slice]
+
+    # Calculate normalisations.
+    normalisations = np.ones_like(weight_sums)
+    normalisations *= tgt_mask
```
Comment (HGWright) [inline lib/iris/analysis/_area_weighted.py:666]: We believe this can be simplified using something like `normalisations = tgt_mask.copy`.

**Review by HGWright — CHANGES_REQUESTED** [review]
Fantastic performance benefits, great work. Just a few minor changes.

File: lib/iris/analysis/_area_weighted.py [hunk 3]
```diff
@@ -1150,3 +459,266 @@ def regrid_callback(*args, **kwargs):
...
+    # Use input cube dtype or convert values to the smallest possible float
+    # dtype when necessary.
+    dtype = np.promote_types(data.dtype, np.float16)
+
+    # Perform regridding on unmasked data.
+    result = _standard_regrid_no_masks(
+        ma.filled(data, 0.0), weights, tgt_shape
+    )
+    # Apply normalisations and masks to the regridded data.
+    result = result * normalisations
```
[the remark this answers is not in the document]
Comment (stephenworsley) [reply lib/iris/analysis/_area_weighted.py:688]: It turns out that the behaviour of `x *= y` is slightly different to `x = x * y`, see https://stackoverflow.com/questions/35496704/numpy-not-working

[the remark this answers is not in the document]
Comment (HGWright) [reply lib/iris/analysis/_area_weighted.py:690]: Oh i didn't spot that difference. Good catch and therefore given the multiplication from `array` to a `masked array` the fact that *= in an inplace action that wouldn't work.

**Review by HGWright — APPROVED** [review]
Great work @stephenworsley, merging this and we will add a whatsnew entry in a seperate PR

**Pull request merged**